keylime-agent/src/keys_handler.rs

```diff
@@ -94,54 +127,119 @@ pub async fn u_key(
-
-        if let Some(payload) = &body.payload {
-            let encr_payload =
-                base64::decode(payload).map_err(Error::from)?;
-            global_encr_payload.extend(encr_payload.iter());
+        Ok(k) => k,
+        Err(e) => {
+            warn!(
+                    "POST u_key returning 400 response. Invalid base64 encoding in encrypted_key: {}", e
+                );
+            return HttpResponse::BadRequest().json(JsonWrapper::error(
+                400,
+                format!("Invalid base64 encoding in encrypted_key: {}", e),
+            ));
+        }
+    };
+
+    // Uses NK (key for encrypting data from verifier or tenant to agent in transit) to
+    // decrypt U and V keys, which will be combined into one key that can decrypt the
+    // payload.
+    //
+    // Reference:
+    // https://github.com/keylime/keylime/blob/f3c31b411dd3dd971fd9d614a39a150655c6797c/ \
+    // keylime/crypto.py#L118
+    let decrypted_key = match crypto::rsa_oaep_decrypt(
+        &quote_data.priv_key,
+        &encrypted_key,
+    )
+    .map_err(Error::from)
+    {
+        Ok(k) => k,
+        Err(e) => {
+            warn!(
+                        "POST u_key returning 400 response. Failed to decrypt encrypted_key: {}", e
+                        );
+            return HttpResponse::BadRequest().json(JsonWrapper::error(
+                400,
+                format!("Failed to decrypt encrypted_key: {}", e),
+            ));
+        }
+    };
+
+    let decrypted_key: SymmKey = match decrypted_key.as_slice().try_into() {
+        Ok(k) => k,
+        Err(e) => {
+            warn!(
+                "POST u_key returning 400 response. Decrypted key length is invalid: {}", e
+                );
+            return HttpResponse::BadRequest().json(JsonWrapper::error(
+                400,
+                format!("Decrypted key length is invalid: {}", e),
+            ));
         }
+    };
 
-        if let Some(symm_key) = try_combine_keys(
-            &mut global_current_keyset,
-            &mut global_other_keyset,
-            quote_data.agent_uuid.as_bytes(),
-            &global_auth_tag,
-        )? {
-            let _ = global_symm_key.replace(symm_key);
-            quote_data.payload_symm_key_cvar.notify_one();
+    let auth_tag = match hex::decode(&body.auth_tag).map_err(Error::from) {
+        Ok(t) => t,
+        Err(e) => {
+            warn!(
+                "POST u_key returning 400 response. Invalid hex encoding in auth_tag: {}", e
+                );
+            return HttpResponse::BadRequest().json(JsonWrapper::error(
+                400,
+                format!("Invalid hex encoding in auth_tag: {}", e),
+            ));
         }
+    };
+
+    let auth_tag = match auth_tag.as_slice().try_into() {
+        Ok(t) => t,
+        Err(e) => {
+            warn!(
+                "POST u_key returning 400 response. Auth tag length is invalid: {}", e
```

Comment:
The length error is already in `e` if I read the code correctly. Maybe later there are other cases identified of why the conversion cannot be done.

Reply:
I did not review all the cases, but modified some.